### PR TITLE
[REG-1701] Record results during replay

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Now records data and uploads for any replay session
- Fixes a bug where FPS data had been broken in a prior PR
- Slightly improves the performance of our RGDebug log statements
- Replay performance was very poor while recording, so this PR also takes some big steps to reduce Json serialization time
  - Slightly optimizes some string to Json conversions
  - Fixes a bug where decimal/float/double json conversion was incorrect for negative values and also improves its performance
  - Uses a single stringbuilder for each tick to avoid tons of string creations and GCs

Performance Before
JSON_BEFORE

Performance After
JSON_AFTER

Performance Facts (for the specific BossRoom frame on the char select screen I used for apples/apples testing)
- JSON Serialization time = 16ms

Of this time , the biggest chunks are spent in these 2 areas.. note that these values have some overlap as many behaviours have floats in them.
- floats - 5.6ms over 3546 invocations
- BehaviourStates - 7ms over 277 invocations

Ultimately, it still takes too long, but the game doesn't freeze on each recorded tick during replay anymore.  Taking the screenshot also takes a very long time as we have to wait for the final render step in the frame and then read the pixel buffer completely before moving on.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextJsonConverter.cs [hunk 1]
```diff
@@ -1,12 +1,52 @@
...
+            stringBuilder.Append(",\"font\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,val.font.name);
+            stringBuilder.Append(",\"fontStyle\":\"");
+            stringBuilder.Append(val.fontStyle.ToString());
+            stringBuilder.Append("\",\"fontSize\":");
+            FloatJsonConverter.WriteToStringBuilder(stringBuilder, val.fontSize);
+            stringBuilder.Append(",\"color\":");
+            ColorJsonConverter.WriteToStringBuilder(stringBuilder, val.color);
+            stringBuilder.Append(",\"raycastTarget\":");
+            stringBuilder.Append((val.raycastTarget ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextJsonConverter.cs:39]:
```suggestion
            stringBuilder.Append(val.raycastTarget ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextMeshProJsonConverter.cs [hunk 2]
```diff
@@ -1,12 +1,54 @@
...
+            stringBuilder.Append(",\"font\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,val.font.name);
+            stringBuilder.Append(",\"fontStyle\":\"");
+            stringBuilder.Append(val.fontStyle.ToString());
+            stringBuilder.Append("\",\"fontSize\":");
+            FloatJsonConverter.WriteToStringBuilder(stringBuilder, val.fontSize);
+            stringBuilder.Append(",\"color\":");
+            ColorJsonConverter.WriteToStringBuilder(stringBuilder, val.color);
+            stringBuilder.Append(",\"raycastTarget\":");
+            stringBuilder.Append((val.raycastTarget ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextMeshProJsonConverter.cs:41]:
```suggestion
            stringBuilder.Append(val.raycastTarget ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextMeshProUGUIJsonConverter.cs [hunk 3]
```diff
@@ -1,12 +1,54 @@
...
+            stringBuilder.Append(",\"font\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,val.font.name);
+            stringBuilder.Append(",\"fontStyle\":\"");
+            stringBuilder.Append(val.fontStyle.ToString());
+            stringBuilder.Append("\",\"fontSize\":");
+            FloatJsonConverter.WriteToStringBuilder(stringBuilder, val.fontSize);
+            stringBuilder.Append(",\"color\":");
+            ColorJsonConverter.WriteToStringBuilder(stringBuilder, val.color);
+            stringBuilder.Append(",\"raycastTarget\":");
+            stringBuilder.Append((val.raycastTarget ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/TextMeshProUGUIJsonConverter.cs:41]:
```suggestion
            stringBuilder.Append(val.raycastTarget ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonUtils.cs [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                    args.ErrorContext.Handled = true;
+                }
+                else
+                {
+                    // do nothing anyway.. but useful for debugging which errors happened
+                    args.ErrorContext.Handled = true;
+                }
+            },
+            // state, behaviours, state, field, child field ... so we can basically see the vector values of a field on a behaviour on an object, but stop there
+            //MaxDepth = 5
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonUtils.cs:170]:
```suggestion
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonUtils.cs [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+    internal class JsonConverterContractResolver : DefaultContractResolver
+    {
+        public static readonly JsonConverterContractResolver Instance = new();
+
+        [CanBeNull]
+        public static JsonConverter GetConverterForType(Type objectType)
+        {
+            JsonConverter converter = null;
+            if (objectType == typeof(float) || objectType == typeof(Single))
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonUtils.cs:182]: TIL about `Single`

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardInputActionObserver.cs [hunk 6]
```diff
@@ -48,14 +50,30 @@ public class KeyboardInputActionData
...
+            DoubleJsonConverter.WriteToStringBuilder(stringBuilder, startTime);
+            stringBuilder.Append(",\"action\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,action);
+            stringBuilder.Append(",\"binding\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,binding);
+            stringBuilder.Append(",\"endTime\":");
+            JsonUtils.EscapeJsonStringIntoStringBuilder(stringBuilder,action);
+            DoubleJsonConverter.WriteToStringBuilderNullable(stringBuilder, endTime);
+            stringBuilder.Append(",\"isPressed\":");
+            stringBuilder.Append((isPressed ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardInputActionObserver.cs:68]:
```suggestion
            stringBuilder.Append(isPressed ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs [hunk 7]
```diff
@@ -78,19 +79,47 @@ public bool ButtonStatesEqual(object obj)
...
+        public void WriteToStringBuilder(StringBuilder stringBuilder)
+        {
+            stringBuilder.Append("{\"startTime\":");
+            DoubleJsonConverter.WriteToStringBuilder(stringBuilder, startTime);
+            stringBuilder.Append(",\"position\":");
+            VectorIntJsonConverter.WriteToStringBuilder(stringBuilder, position);
+            stringBuilder.Append(",\"worldPosition\":");
+            VectorJsonConverter.WriteToStringBuilderVector3Nullable(stringBuilder, worldPosition);
+            stringBuilder.Append(",\"leftButton\":");
+            stringBuilder.Append( (leftButton ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs:94]:
```suggestion
            stringBuilder.Append(leftButton ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs [hunk 8]
```diff
@@ -78,19 +79,47 @@ public bool ButtonStatesEqual(object obj)
...
+            stringBuilder.Append(",\"middleButton\":");
+            stringBuilder.Append( (middleButton ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs:96]:
```suggestion
            stringBuilder.Append(middleButton ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs [hunk 9]
```diff
@@ -78,19 +79,47 @@ public bool ButtonStatesEqual(object obj)
...
+            stringBuilder.Append(",\"rightButton\":");
+            stringBuilder.Append( (rightButton ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs:98]:
```suggestion
            stringBuilder.Append(rightButton ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs [hunk 10]
```diff
@@ -78,19 +79,47 @@ public bool ButtonStatesEqual(object obj)
...
+            stringBuilder.Append(",\"forwardButton\":");
+            stringBuilder.Append( (forwardButton ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs:100]:
```suggestion
            stringBuilder.Append(forwardButton ? "true" : "false");
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs [hunk 11]
```diff
@@ -78,19 +79,47 @@ public bool ButtonStatesEqual(object obj)
...
+            stringBuilder.Append(",\"backButton\":");
+            stringBuilder.Append( (backButton ? "true" : "false"));
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/MouseInputActionObserver.cs:102]:
```suggestion
            stringBuilder.Append(backButton ? "true" : "false");
```

**Review by addisonbgross — APPROVED** [review]
Look good. I have a whole bunch of tiny comments and nits. I may be wrong about some bracket usage 😅

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/InGameObjectFinder.cs [hunk 12]
```diff
@@ -479,7 +477,7 @@ private RecordedGameObjectState CreateStateForTransform(List<RecordedGameObjectS
             return null;
         }
 
-        // pre-allocate this rather large list 1 time to avoid memory stuff on each tick
+        // pre-allocate this rather large list 1 time to avoid memory stuff on each stick
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/InGameObjectFinder.cs:480]: ... sigh... will fix

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/DecimalJsonConverter.cs [hunk 13]
```diff
@@ -1,26 +1,135 @@
...
+            {
+                stringBuilder.Append("00000");
+            }
+            else
+            {
+                stringBuilder.Append("000000");
+            }
+
+            stringBuilder.Append(value.ToString(NumberFormatInfo));
+        }
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/DecimalJsonConverter.cs:116]: it is a prepend.. . I append the needed 0s before appending the number

this method is handling the right side of the decimal portion of a floating point number 

i get that right side value as an up to 7 digit int by `var remainder = (int)((val % 1) * 10_000_000)`

the issue is that `.0000007` will be `7` for that `remainder` and needs to have 6 zeros which is where this append 0s before the remainder value is used ... thus effectively prepending the 0s onto the remainder in the final string result

[the remark this answers is not in the document]
Comment (addisonbgross) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/JsonConverters/DecimalJsonConverter.cs:116]: ah I read things out of order. Makes sense 👍🏻

**Review by vontell — APPROVED** [review]
Thanks for working on these changes, this seemed like a lot of work... but will be great in user conversations where we can mention that we've spent a good amount of time writing custom code to make things performant.

**Pull request merged**